R/civis_future.R

```diff
@@ -0,0 +1,167 @@
+#' Evaluate an expression in Civis Platform
+#'
+#' @param ... Arguments to \code{\link{CivisFuture}}
+#' @return The result of evaluating \code{expr}.
+#'
+#' @examples \dontrun{
+#'  eval_civis(quote(sqrt(10)))
+#' }
+#'
+#' @export
+civis_platform <- function(...) {
+  future <- CivisFuture(...)
+  if (!future$lazy) future <- run(future)
+  future
+}
+class(civis_platform) <- c("CivisFuture", "future", "function")
+
+#' Evaluate an expression in Civis Platform
+#' @inheritParams future::Future
+#' @param required_resources resources
+#' @param docker_image_name the image for the container script.
+#' @param docker_image_tag the tag for the Docker image.
+#' @param ... arguments to \code{\link{scripts_post_containers}}
+#'
+#' @return A future evaluating \code{expr} on the given container.
+#'
+#' @export
+CivisFuture <- function(expr = NULL,
+                        envir = parent.frame(),
+                        substitute = FALSE,
+                        globals = TRUE,
+                        packages = NULL,
+                        lazy = FALSE,
+                        local = TRUE,
+                        gc = FALSE,
+                        earlySignal = FALSE,
+                        label = NULL,
+                        required_resources = list(cpu = 1024, memory = 2048, diskSpace = 4),
+                        docker_image_name = "civisanalytics/datascience-r",
+                        docker_image_tag = "2.0.0",
+                         ...) {
+
+  # import_future imports the function from future without :::, which triggers R CMD CHECK
+  getGlobalsAndPackages <- import_future("getGlobalsAndPackages")
+  gp <- getGlobalsAndPackages(expr, envir = envir, globals = globals)
+
+  ## if there are globals, assign them in envir
+  env <- new.env()
+  if (length(gp) > 0) {
+    env <- list2env(gp$globals)
+
+  }
+
+  future <- future::Future(expr = expr,
+                           envir = env,
+                           substitute = substitute,
+                           globals = gp$globals,
+                           packages = unique(c(packages, gp$packages)),
+                           lazy = lazy,
+                           local = local,
+                           gc = gc,
+                           earlySignal = earlySignal,
+                           label = label,
+
+                           # extra args to scripts_post_containers
+                           required_resources = required_resources,
+                           docker_image_name = docker_image_name,
+                           docker_image_tag = docker_image_tag, ...)
+  structure(future, class = c("CivisFuture", class(future)))
+}
+
+#' @importFrom future run value resolved
+NULL
+
+#' @export
+run.CivisFuture <- function(future, ...) {
+  cargo <- c(expr = future$expr, envir = future$envir, packages = list(future$packages))
```

Comment:
is the package `future` in `list(future$packages)` ?  If so, I wonder if we could strip it out rather than relying on suppressing warnings in the `r_remote_eval.R` script.

Reply:
Now that I think about it, we now have to depend on future for the API. `future` will automatically get loaded with `library(civis)`, so no warning.